File: Week6 (Graph [SSSP, APSP, BFS, DFS])/Single Source Shortest Path (SSSP)/cmy-sssp-bellman-ford.cpp
#include <iostream>
#include <vector>
#include <climits>

using namespace std;

bool bellman_ford(int source, const vector<tuple<int, int, int>>& edges, vector<int>& distance) {
    int n = distance.size();
    fill(distance.begin(), distance.end(), INT_MAX);
    distance[source] = 0;
    
    for (int i = 1; i < n - 1; ++i) {
        for (const auto& edge : edges) {
            int u, v, w;
            tie(u, v, w) = edge;
            if (distance[u] != INT_MAX && distance[u] + w < distance[v]) {
                distance[v] = distance[u] + w;
            }
        }
    }
    
    for (const auto& edge : edges) {
        int u, v, w;
        tie(u, v, w) = edge;
        if (distance[u] != INT_MAX && distance[u] + w < distance[v]) {
            return false; // Negative weight cycle detected
        }
    }
    
    return true; // No negative weight cycle
}

int main() {
    int n, m;
    cin >> n >> m; // Number of nodes and edges
    
    vector<tuple<int, int, int>> edges;
    for (int i = 0; i < m; ++i) {
        int u, v, w;
        cin >> u >> v >> w; // Edge from u to v with weight w
        edges.emplace_back(u, v, w);
    }
    
    vector<int> distance(n + 1);
    if (bellman_ford(1, edges, distance)) {
        for (int i = 1; i <= n; ++i) {
            if (distance[i] == INT_MAX) {
                cout << "INF ";
            } else {
                cout << distance[i] << " ";
            }
        }
        cout << endl;
    } else {
        cout << "Negative weight cycle detected!" << endl;
    }
    
    return 0;
}

File: Week6 (Graph [SSSP, APSP, BFS, DFS])/Single Source Shortest Path (SSSP)/cmy-sssp-bellman-ford.py
import sys

def bellman_ford(source, edges, n):
    distance = [float('inf')] * (n + 1)
    distance[source] = 0
    
    for _ in range(n - 1):
        for u, v, weight in edges:
            if distance[u] != float('inf') and distance[u] + weight < distance[v]:
                distance[v] = distance[u] + weight
    
    # Check for negative weight cycles
    for u, v, weight in edges:
        if distance[u] != float('inf') and distance[u] + weight < distance[v]:
            return None  # Negative weight cycle detected
    
    return distance

def main():
    input = sys.stdin.read
    data = input().split()
    
    n = int(data[0])
    m = int(data[1])
    
    edges = []
    index = 2
    for _ in range(m):
        u = int(data[index])
        v = int(data[index + 1])
        w = int(data[index + 2])
        edges.append((u, v, w))
        index += 3
    
    distance = bellman_ford(1, edges, n)
    
    if distance is None:
        print("Negative weight cycle detected!")
    else:
        for i in range(1, n + 1):
            if distance[i] == float('inf'):
                print("INF", end=" ")
            else:
                print(distance[i], end=" ")
        print()

if __name__ == "__main__":
    main()
